### ISSUE TITLE
mkosi has switched to systemd-repart

### ISSUE THREAD

**Issue by behrmann** [systemd/mkosi#1290]
Not an issue, but as a heads up to users of the `main` branch. mkosi has switched to systemd-repart in #1276.

Please have a look at [the news file](https://github.com/systemd/mkosi/blob/381a23fbd0ba1b33e9e6a04cc8a81aacde65b6b8/NEWS.md#v15) how to adapt your configs and mind that for now you will need a current version of repart built from the systemd sources, using
```ini
ExtraSearchPaths=/path/to/the/systemd/builddir
```

**Comment by benschlueter** [comment]
The quickfix is not working if you specify partitions on your own with `mkosi.repart`

**Comment by DaanDeMeyer** [comment]
> The quickfix is not working if you specify partitions on your own with mkosi.repart

Please open a new issue with logs and a detailed description of the problem and I'll take a look

**Comment by benschlueter** [comment]
Not sure what was causing it. However, the error is now gone. Before I received `systemd-repart: unrecognized option '--exclude-partitions=esp,xbootldr'` indicating that the wrong `systemd-repart` binary is used, but I specified the right one in the config and did not receive the `systemd-repart: unrecognized option '--defer-partitions'` error. 

Nonetheless, the qcow2 image is not built correctly (not sure if it is mkosi or if I am not able to specify the partitions correctly :smile: ). The resulting image is 1MiB in size which isn't correct.

config
```
[Distribution]
Distribution=arch

[Output]
Format=disk
Bootable=yes
QCow2=yes
Output=image.qcow2
WorkspaceDirectory=/tmp
ExtraSearchPaths=/home/bschlueter/University/Github/systemd/build

[Packages]
Packages=
         openssh
         vim
         crictl
         critest
         kubeadm
         kubectl
         socat
         kubelet
         ethtool
         containerd

[Validation]
Password=password

[Content]
Cache=.cache
Autologin=yes

[Host]
QemuHeadless=yes
```

root.conf
```
[Partition]
Type=root
Label=root
Format=xfs
SizeMinBytes=10G
SizeMaxBytes=10G
```

esp.conf
```
[Partition]
Type=esp
Label=esp
SizeMinBytes=256M
SizeMaxBytes=256M
```

tail - build output
```
‣  Unmounting image…
No machine ID set, using randomized partition UUIDs.
Automatically determined minimal disk image size as 10.2G.
Sized '/tmp/bwgs_trs/staging/image.qcow2' to 10.2G.
Applying changes.
Wiped block device.
Discarded entire block device.
Successfully wiped file system signatures from future partition 1.
Formatting future partition 1.
/dev/loop0 successfully formatted as xfs (label "root", uuid 53b7bfa9-602b-4ab6-86a0-f87d4239be1a)
Successfully formatted future partition 1.
Adding new partition 1 to partition table.
Writing new partition table.
All done.
‣  Generating combined kernel + initrd boot file…
No machine ID set, using randomized partition UUIDs.
Automatically determined minimal disk image size as 10.2G, current image size is 10.2G.
File '/tmp/bwgs_trs/staging/image.qcow2' already is of requested size or larger, not growing. (10.2G >= 10.2G)
Applying changes.
Successfully wiped file system signatures from future partition 0.
Successfully discarded data from future partition 0.
Adding new partition 0 to partition table.
Writing new partition table.
All done.
‣ Converting image file to qcow2…
‣ Saving manifest image.manifest
‣ Digging holes into output file /home/bschlueter/University/Github/mkosi/images/image.qcow2…
‣ Resulting image size is 1.0M, consumes 188.0K.
```

Do you have any mwe's for repart which I can use as a basis?

**Comment by rezib** [comment]
This is a great feature, I love it for sure and thank you for this. But do you plan to maintain bugfix releases of v14 for users not running bleeding edge version of systemd? I am encountering bugs in v14 that are fixed in main (eg. build of rocky9 images) but I cannot run the main branch and future v15 on my production servers anymore. Do you have any plan regarding this?

**Comment by behrmann** [comment]
I'm not sure we have the manpower to support a v14 branch. Personally, if you want to run the main branch, I'd go with having a local checkout of the systemd repo and building the needed tools from source as we do in CI, see [action.yaml](https://github.com/systemd/mkosi/blob/main/action.yaml). This has the added benefit of users not having to wait for new systemd to arrive in their distro (and them having to put any work to switching the systemd they run to something like a backports version) und us being able to use newer things right away.

**Comment by rezib** [comment]
Thank you @behrmann for your answer. FWIW, I successfully tested this approach on Debian bullseye by building a custom `systemd-mkosi` package with the executables mentionned in [action.yaml](https://github.com/systemd/mkosi/blob/main/action.yaml). As a side note, this requires backporting systemd's `ukify` to Python 3.9 as it expects Python >= 3.10. Anyone feel free to contact me if you are interested in this `systemd-mkosi` Debian bullseye binary (or source) package.

**Comment by DaanDeMeyer** [comment]
@Simon-L We're planning to release v15 with a strict requirement on v254 (since we'll be relying on new features from kernel-install very soon). We're not going to delay that release until systemd 254 is available everywhere.

We simply don't have the manpower to maintain support for v14. It's only me writing most of the code with support from @behrmann and I'm stretched thin as it is getting mkosi in a good state for the next release.

I am going to try and add a backwards compatible guarantee from v15 onwards to make sure we don't get into a situation like this again.

**Comment by pbjhelmert** [comment]
Is the v15 release blocked on anything in particular? It seems that CentOS-derivative >=9 builds with v14 are broken due to the PowerTools -> CRB change:
```
‣     Mounting API VFS…
/usr/lib/python3/dist-packages/dnf/const.py:22: DeprecationWarning: The distutils package is deprecated and slated for removal in Python 3.12. Use setuptools or check PEP 632 for potential alternatives
  import distutils.sysconfig
/usr/lib/python3/dist-packages/dnf/const.py:22: DeprecationWarning: The distutils.sysconfig module is deprecated, use sysconfig instead
  import distutils.sysconfig
AppStream                                                                                                                                                                         540 kB/s | 7.1 MB     00:13
BaseOS                                                                                                                                                                            2.8 MB/s | 1.9 MB     00:00
extras                                                                                                                                                                             23 kB/s |  10 kB     00:00
PowerTools                                                                                                                                                                         19 kB/s |  20 kB     00:01
Errors during downloading metadata for repository 'PowerTools':
  - Status code: 404 for https://mirrors.rockylinux.org/mirrorlist?arch=x86_64&repo=PowerTools-9 (IP: 199.232.194.132)
  - Status code: 404 for https://mirrors.rockylinux.org/mirrorlist?arch=x86_64&repo=PowerTools-9 (IP: 199.232.198.132)
Error: Failed to download metadata for repo 'PowerTools': Cannot prepare internal mirrorlist: Status code: 404 for https://mirrors.rockylinux.org/mirrorlist?arch=x86_64&repo=PowerTools-9 (IP: 199.232.198.132)
Traceback (most recent call last):
```

**Comment by keszybz** [comment]
We could have a v14-stable branch with some point releases. Even if v15 is released, it wouldn't be suitable for updates in stable releases, so it'd make sense to keep both versions updated in parallel for a bit. But as Daan said, the people currently working on this don't have the resources to provide this. Somebody else would need to volunteer. (If there's a e.g. specific list of commits that could be backported, that'd be a good start.)

**Comment by pbjhelmert** [comment]
That sounds great, thanks!
For me, the primary fix is fee9cdf which I semi-manually backported here: https://github.com/pbjhelmert/mkosi/commit/1f51216ac9fce7d4af05fbd70158a43ed3784205

**Comment by septatrix** [comment]
> @Simon-L We're planning to release v15 with a strict requirement on v254 (since we'll be relying on new features from kernel-install very soon). We're not going to delay that release until systemd 254 is available everywhere.

Is v254 now required for v15? There is no mention of it in the release notes...

**Comment by behrmann** [comment]
It's not in the news - where it probably should be, too - but in the [requirements section of the docs](https://github.com/systemd/mkosi/blob/main/mkosi/resources/mkosi.md#requirements)
> mkosi currently requires systemd 254 to build bootable disk images.

**Comment by septatrix** [comment]
What is the equivalent of `RootSize=...` now? When I try to add partition config files I always get either empty partitions or permission errors for not being able to mount loopback devices (or does that still require root against the statement that mkosi now doesn't require root)?

It's certainly nice having the whole arsenal of repart options available and I will definitely try to use it to configure A/B and a split rw /var partition down the line though I am quite new to repart and do not understand what I might do wrong after reading the man pages

**Comment by DaanDeMeyer** [comment]
You can use the source code as a starting point. I'll add the definitions from the mkosi source to the docs to make this easier.

**Comment by DaanDeMeyer** [comment]
@septatrix PR documenting the defaults: https://github.com/systemd/mkosi/pull/1817

**Comment by septatrix** [comment]
The documentation still says:

> `ImageId=`, `--image-id=`
> [...]. If this option is used the root, /usr/ and Verity partitions in the image will have their labels set to this (possibly suffixed by the image version).

This seems to no longer be the case after this switch. Can someone else confirm this?

Sadly I do not think there is a way to get this behaviour back with the current repart capabilities (at least not automatically from within mkosi). As a workaround I have now added `Label=%M_%A{,_verity,_verity_sig}` to most of my repart files.

**Comment by DaanDeMeyer** [comment]
Yeah this seems out of date, will fix the docs.
